From b38a669ecb043d601743635f2fbcfa5cde4f5b1e Mon Sep 17 00:00:00 2001
From: Dave Mankoff <mankoff@google.com>
Date: Wed, 11 Sep 2019 14:14:45 -0400
Subject: [PATCH 015/297] Ensure that the Application is constructed before any
 Service.

This fixes dependency injection issues and brings Service construction
in line with the way that Activities are created. Specifically, this
ensures that there is always a reference to a Context available at
the time a Service is constructed, such that the Context can then be
injected into the constructor of the Service.

Bug: 140868981
Test: atest SystemUITests
Change-Id: If85e4106fef8c2fd12f7ca9e8b7fd686af62e29b
(cherry picked from commit 66d22c8a59836b2e3f22074e358696275ed10ab7)
---
 core/java/android/app/ActivityThread.java | 18 ++++--------------
 1 file changed, 4 insertions(+), 14 deletions(-)

diff --git a/core/java/android/app/ActivityThread.java b/core/java/android/app/ActivityThread.java
index a6784780d72..1cc849936fd 100644
--- a/core/java/android/app/ActivityThread.java
+++ b/core/java/android/app/ActivityThread.java
@@ -3982,25 +3982,15 @@ public final class ActivityThread extends ClientTransactionHandler {
         LoadedApk packageInfo = getPackageInfoNoCheck(
                 data.info.applicationInfo, data.compatInfo);
         Service service = null;
-        try {
-            java.lang.ClassLoader cl = packageInfo.getClassLoader();
-            service = packageInfo.getAppFactory()
-                    .instantiateService(cl, data.info.name, data.intent);
-        } catch (Exception e) {
-            if (!mInstrumentation.onException(service, e)) {
-                throw new RuntimeException(
-                    "Unable to instantiate service " + data.info.name
-                    + ": " + e.toString(), e);
-            }
-        }
-
         try {
             if (localLOGV) Slog.v(TAG, "Creating service " + data.info.name);
 
             ContextImpl context = ContextImpl.createAppContext(this, packageInfo);
-            context.setOuterContext(service);
-
             Application app = packageInfo.makeApplication(false, mInstrumentation);
+            java.lang.ClassLoader cl = packageInfo.getClassLoader();
+            service = packageInfo.getAppFactory()
+                    .instantiateService(cl, data.info.name, data.intent);
+            context.setOuterContext(service);
             service.attach(context, this, data.info.name, data.token, app,
                     ActivityManager.getService());
             service.onCreate();
-- 
2.35.1

